pandas/core/groupby/groupby.py

```diff
@@ -807,7 +807,10 @@ def _try_cast(self, result, obj, numeric_only: bool = False):
             dtype = obj.dtype
 
         if not is_scalar(result):
-            if is_extension_array_dtype(dtype) and dtype.kind != "M":
+            if (
+                is_extension_array_dtype(dtype)
+                and dtype.kind != "M"
+            ):
```

Comment:
up until a moment ago this had a length check; without that check this looks like it isnt actually changing anything

Reply:
in the previous commit, i did add a length check, but i do not feel it was right solution. the issue is `dtype` preserves its categories @jbrockmendel

Reply:
so `dtype.categories` preserves the original value, in this case, [0, 1] as its categories, however, after aggregation, we should NOT use the original categories, but the new one based on the result